### ISSUE TITLE
Make IdentityProvider.EntityId immutable after added to identityprovider collection

### ISSUE THREAD

**Issue by AndersAbel** [Sustainsys/Saml2#819]
Add an `internal` flag that makes the EntityId of an IdentityProvder immutable once it's been added to the IdentityProvider collection to remedy the scenario commented in #540.

**Comment by explunit** [comment]
IdentityProvider.EntityId is already a private setter. Do you mean the ID property of the EntityId object (System.IdentityModel.Metadata)?

```
            var entityId = new EntityId("http://idp.example.com");
            var idp = new IdentityProvider(entityId, StubFactory.CreateSPOptions());

            subject.Add(idp);
            //this doesn't compile:
            //idp.EntityId = new EntityId("http://idp.example.com");

            // but this changes the ID
            idp.EntityId.Id = "http://idp2.example.com";
```

This probably should be implemented with #637 in mind

**Comment by AndersAbel** [comment]
@explunit You're right. I obviously already did what I can. #637 ill definitely be relevant for this, but it still has the same problem: The class is not immutable.

The basic problem is that a property of the IdentityProvider is used as the key to the dictionary it is stored in. If you change the property after it's been added to the dictionary they will get out of sync.

I don't think there's any elegant way around this. Leaving this open for now if anyone has any ideas.
